contracts/types/YieldSplitter.sol

```diff
@@ -0,0 +1,141 @@
+        uint256 id;
+        address depositor;
+        uint256 principalAmount; // Total amount of sOhm deposited as principal, 9 decimals.
+        uint256 agnosticAmount; // Total amount deposited priced in gOhm. 18 decimals.
+    }
+
+    uint256 public idCount;
+    mapping(uint256 => DepositInfo) public depositInfo; // depositId -> DepositInfo
+    mapping(address => uint256[]) public depositorIds; // address -> Array of the deposit id's deposited by user
+
+    /**
+        @notice Constructor
+        @param gOHM_ Address of gOHM.
+    */
+    constructor(address gOHM_) {
+        gOHM = gOHM_;
+    }
+
+    /**
+        @notice Create a deposit.
+        @param depositor_ Address of depositor
+        @param amount_ Amount in gOhm. 18 decimals.
+    */
+    function _deposit(
+        address depositor_,
+        uint256 amount_
+    ) internal returns (uint256 depositId) {
+        depositorIds[depositor_].push(idCount);
+
+        depositInfo[idCount] = DepositInfo({
+            id: idCount,
+            depositor: depositor_,
+            principalAmount: IgOHM(gOHM).balanceFrom(amount_),
+            agnosticAmount: amount_
+        });
+
+        depositId = idCount;
+        idCount++;
+    }
+
+    /**
+        @notice Add more gOhm to the depositor's principal deposit.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOhm to add. 18 decimals.
+    */
+    function _addToDeposit(uint256 id_, uint256 amount_) internal {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        userDeposit.principalAmount += IgOHM(gOHM).balanceFrom(amount_);
+        userDeposit.agnosticAmount += amount_;
+    }
+
+    /**
+        @notice Withdraw part of the principal amount deposited.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOHM to withdraw.
+    */
+    function _withdrawPrincipal(uint256 id_, uint256 amount_) internal {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        userDeposit.principalAmount -= IgOHM(gOHM).balanceFrom(amount_); // Reverts if amount > principal due to underflow
+        userDeposit.agnosticAmount -= amount_;
+    }
+
+    /**
+        @notice Withdraw all of the principal amount deposited.
+        @param id_ Id of the deposit.
+        @return amountWithdrawn : amount of gOHM withdrawn. 18 decimals.
+    */
+    function _withdrawAllPrincipal(uint256 id_) internal returns (uint256 amountWithdrawn) {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        amountWithdrawn = IgOHM(gOHM).balanceTo(userDeposit.principalAmount);
+        userDeposit.principalAmount = 0;
+        userDeposit.agnosticAmount -= amountWithdrawn;
+    }
+
+    /**
+        @notice Redeem excess yield from your deposit in sOHM.
+        @param id_ Id of the deposit.
+        @return amountRedeemed : amount of yield redeemed in gOHM. 18 decimals.
+    */
+    function _redeemYield(uint256 id_) internal returns (uint256 amountRedeemed) {
```

Comment:
would there be any use cases of partial redemptions to include in a separate function?

Reply:
Can't think of any use case but it might be one in future.

From version maintenance POV i think its ok for us to modify this base contract by adding functions. But probably less ok if we modify existing functions. Because if we redeploy an existing implementation with new version and it has an extra internal functions its still fine but deploying with modified base contract function will produce diff results.

So im ok holding off adding it until we need it.